service/migration_manager.cc

```diff
@@ -1003,6 +1004,71 @@ future<> migration_manager::announce<schema_change>(std::vector<mutation> schema
 template
 future<> migration_manager::announce<topology_change>(std::vector<mutation> schema, group0_guard, std::string_view description);
 
+future<schema_ptr> migration_manager::inject_internal_columns(schema_ptr target, 
+        const internal_column_set& column_set)
+{
+    const auto find_schema = [table_id=target->id()](migration_manager& mm) {
+        return mm._storage_proxy.get_db().local().find_schema(table_id);
+    };
+    const auto columns_exist = [&column_set](const schema& s) {
+        const auto& non_empty = column_set.regular_columns.empty()
+            ? column_set.static_columns
+            : column_set.regular_columns;
+        return s.get_column_definition(non_empty.begin()->name) != nullptr;
+    };
+
+    if (columns_exist(*target)) {
+        co_return target;
+    }
+    mlogger.info("Injecting internal column set \"{}\" into the table \"{}.{}\"", 
+        to_string_view(column_set.prefix), target->ks_name(), target->cf_name());
+    co_await container().invoke_on(0, coroutine::lambda([&column_set, &find_schema, &columns_exist] (migration_manager& mm) -> future<> {
+        auto retries = mm.get_concurrent_ddl_retries();
+        while (true)  {
+            try {
+                auto guard = co_await mm.start_group0_operation();
+                const auto s = find_schema(mm);
```

Comment:
Repeating my previous comment - everything here would have been clearer if you just had here the direct code that looks up, after taking the lock, the table with the given id, instead of having this lambda find_schema 50 lines up.